akka-stream/src/main/scala/akka/stream/impl/NewLayout.scala

```diff
@@ -0,0 +1,297 @@
+/**
+ * Copyright (C) 2015-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.impl
+
+import akka.stream.impl.StreamLayout.{ AtomicModule, Module }
+import akka.stream.impl.fusing.GraphStages.MaterializedValueSource
+import akka.stream.{ Shape, OutPort, InPort, Attributes }
+
+import scala.collection.immutable
+
+object NewLayout {
+
+  // Materialized values
+  // Attributes
+  // MatValueSource
+  // Fusing
+
+  sealed trait Traversal {
+    def next: Traversal
+
+    def concat(that: Traversal): Traversal = {
+      if (that eq EmptyTraversal) this
+      else Concat(this, that)
+    }
+  }
+
+  final case class PushMaterializedValue(next: Traversal) extends Traversal
+  final case class PopMaterializedValue(next: Traversal) extends Traversal
+  final case class TapMaterializedValue(src: MaterializedValueSource[Any], next: Traversal) extends Traversal
+
+  final case class Concat(first: Traversal, next: Traversal) extends Traversal
+
+  final case class AddAttributes(attributes: Attributes, next: Traversal) extends Traversal
+
+  final case class MaterializeAtomic(module: Module, outToSlots: Array[Int]) extends Traversal {
+    override def next: Traversal = EmptyTraversal
+  }
+
+  object EmptyTraversal extends Traversal {
+    def next: Traversal =
+      throw new IllegalStateException("EmptyTraversal has no next Traversal but next was called.")
+
+    override def concat(that: Traversal): Traversal = that
+  }
+
+  private def initShape(shape: Shape): Unit = {
+    // Initialize port IDs
+    val inIter = shape.inlets.iterator
+    var i = 0
+    while (inIter.hasNext) {
+      inIter.next.id = i
+      i += 1
+    }
+
+    val outIter = shape.outlets.iterator
+    i = 0
+    while (outIter.hasNext) {
+      outIter.next.id = i
+      i += 1
+    }
+  }
+
+  object TraversalBuilder {
+    def atomic(module: AtomicModule): TraversalBuilder = {
+      initShape(module.shape)
+
+      if (module.outPorts.isEmpty) {
+        CompletedTraversalBuilder(
+          traversal = Some(MaterializeAtomic(module, Array.ofDim[Int](module.shape.outlets.size))),
+          inSlots = module.shape.inlets.size,
+          inToOffset = module.shape.inlets.map(in ⇒ in → in.id).toMap
+        )
+      } else {
+        AtomicTraversalBuilder(
+          module,
+          Array.ofDim[Int](module.shape.outlets.size),
+          module.shape.outlets.size
+        )
+      }
+    }
+  }
+
+  trait TraversalBuilder {
```

Comment:
I found it very helpful—also for myself—to just note down a few bullet points describing some ideas or intentions, like [for ActorSystemImpl](https://github.com/akka/akka/pull/21128/files#diff-8dfc3787ca99b3f3f62a159c8565ab0bR40), even when things are not at all finished. On the contrary, I mostly start with those comments before writing any code at all. These comments have neither the quality nor the expected longevity of “documentation”.

But we can of course also discuss in a hangout next week.